starlink_grpc.py

```diff
@@ -368,6 +375,145 @@
+    "seconds_obstructed": float,
+    "obstruction_duration": float,
+    "obstruction_interval": float,
+    "direction_azimuth": float,
+    "direction_elevation": float,
+    "is_snr_above_noise_floor": bool,
+}
+
+_OBSTRUCTION_TYPES = {
+    "wedges_fraction_obstructed[12]": float,
+    "raw_wedges_fraction_obstructed[12]": float,
+    "valid_s": float,
+}
+
+_HISTORY_GENERAL_TYPES = {
+    "samples": int,
+    "end_counter": int,
+}
+
+_HISTORY_BULK_TYPES = {
+    "pop_ping_drop_rate[]": float,
+    "pop_ping_latency_ms[]": float,
+    "downlink_throughput_bps[]": float,
+    "uplink_throughput_bps[]": float,
+    "snr[]": float,
+    "scheduled[]": bool,
+    "obstructed[]": bool,
+}
+
+_PING_DROP_TYPES = {
+    "total_ping_drop": float,
+    "count_full_ping_drop": int,
+    "count_obstructed": int,
+    "total_obstructed_ping_drop": float,
+    "count_full_obstructed_ping_drop": int,
+    "count_unscheduled": int,
+    "total_unscheduled_ping_drop": float,
+    "count_full_unscheduled_ping_drop": int,
+}
+
+_PING_DROP_RL_TYPES = {
+    "init_run_fragment": int,
+    "final_run_fragment": int,
+    "run_seconds[1,61]": int,
+    "run_minutes[1,61]": int,
+}
+
+_PING_LATENCY_TYPES = {
+    "mean_all_ping_latency": float,
+    "deciles_all_ping_latency[11]": float,
+    "mean_full_ping_latency": float,
+    "deciles_full_ping_latency[11]": float,
+    "stdev_full_ping_latency": float,
+}
+
+_LOADED_LATENCY_TYPES = {
+    "load_bucket_samples[15]": int,
+    "load_bucket_min_latency[15]": float,
+    "load_bucket_median_latency[15]": float,
+    "load_bucket_max_latency[15]": float,
+}
+
+_USAGE_TYPES = {
+    "download_usage": int,
+    "upload_usage": int,
+}
+
+_OPTIONAL_FIELDS = {
+    "load_bucket_min_latency[15]", "load_bucket_median_latency[15]", "load_bucket_max_latency[15]",
+    "obstructed[]", "obstruction_duration", "obstruction_interval", "scheduled[]",
+    "seconds_obstructed", "snr", "snr[]"
+}
+
+# Minimal parsing for use below. For a more complete regex, see
+# dish_common.BRACKETS_RE
+_TYPE_HINT_RE = re.compile(r"[^[]*\[(\d+,|)")
+
+
+def _type_hints(types):
+    def xform(items):
+        for key, val in items:
+            seq_match = _TYPE_HINT_RE.match(key)
+            opt_val = Optional[val] if key in _OPTIONAL_FIELDS else val
+            if seq_match is None:
+                yield key, opt_val
+            else:
+                # Field names in returned data dicts do not include sequence
+                # length, so strip it out if present.
+                yield seq_match.group() + "]", Sequence[opt_val]
+
+    return {key: val for key, val in xform(types.items())}
+
+
+def _type_hints_bulk(types):
+    # As a special case, bulk history is all sequences, but does not include
+    # the brackets suffix in the returned data dict keys.
+    return {
+        key[:-2]: Sequence[Optional[val] if key in _OPTIONAL_FIELDS else val]
+        for key, val in types.items()
+    }
+
+
+if _typed_dict_ok:
```

Comment:
Has this been tested in an IDE to validate return types? I've never seen an approach like this before

Reply:
No, I've never used type hints before, which is why I asked you to review. I did verify the types look as expected in the Python interpreter, but if this doesn't look obviously correct to you, then I will find a way to verify with a static type checker prior to merging.

Reply:
So... I ran this through `mypy` and it choked pretty badly. It will only accept a dict literal for the second arg below, which is consistent with PEP 589's (non)requirements for type checkers.

In other words, `TypedDict` cannot be used the way I'm trying to use it. It's valid run-time Python code, but static code checkers will not necessarily be able to use it, which would defeat the purpose of adding type hints.

There's a separate problem with the fallback for Python 3.7, but that's minor in comparison.